server/channels/db/migrations/postgres/000106_threadmemberships_cleanup.up.sql

```diff
@@ -0,0 +1,15 @@
+DELETE FROM threadmemberships WHERE EXISTS (
```

Comment:
I think this deletes all records in the `ThreadMemberships` table:
```
> select count(*) from threadmemberships;
 count
-------
   445
(1 row)

(mmuser@[local]:5432) 15:01:30 [mattermost_test]
> DELETE FROM threadmemberships WHERE EXISTS (
    SELECT
        threadmemberships.*
    FROM
        threadmemberships
        JOIN threads ON threads.PostId = threadmemberships.postid
        LEFT JOIN channelmembers ON channelmembers.userid = threadmemberships.userid
            AND threads.channelid = channelmembers.channelid
        JOIN channels ON channels.id = threads.channelid
    WHERE
        threads.postid = threadmemberships.postid
        AND channelmembers.channelid IS NULL
        AND (Channels.Type != 'D'
            OR Channels.Type != 'G')
);
DELETE 445
```

but I'm definitely a valid member of thread in my localhost:
<img width="872" alt="image" src="https://user-images.githubusercontent.com/1023171/228027533-eee4fc4d-f50b-40c5-9492-1ec816f7625e.png">

Reply:
Thanks @lieut-data, I should have catch it, the correct query should be:

```sql
DELETE FROM threadmemberships USING threads (postid, userid) IN (
    SELECT
        threadmemberships.postid,
        threadmemberships.userid
    FROM
        threadmemberships
        JOIN threads ON threads.postid = threadmemberships.postid
        LEFT JOIN channelmembers ON channelmembers.userid = threadmemberships.userid
            AND threads.channelid = channelmembers.channelid
        JOIN channels ON channels.id = threads.channelid
    WHERE
        channelmembers.channelid IS NULL
        AND Channels.Type != 'D'
        AND Channels.Type != 'G'
);
```

BTW, can we write unit tests for migrations?

Reply:
> BTW, can we write unit tests for migrations?

I'm not sure, but can definitely see how this would be useful!

While we're in here, can you clarify this part. Is it strictly necessary to ignore DMs and GMs? I would have assumed the user to have been a part of those channels with a valid channel membership, so wondering if the check is redundant (and expensive, given the need to `JOIN Channels`).
```sql
        AND Channels.Type != 'D'
        AND Channels.Type != 'G'
```

Reply:
Yep possible. cc @agarciamontoro

Reply:
Can we check with the SRE team if this also happens in Community? If this is bad data coming from the load-test, then we should file a bug. I would have assumed the same as Jesse said, but I'm not familiar with the logic there.